ffi/src/lib.rs

```diff
@@ -730,6 +730,61 @@ pub unsafe extern "C" fn fwd_free_value(value: *const Value) {
     }
 }
 
+/// Struct returned by `fwd_create_db` and `fwd_open_db`
+#[derive(Debug)]
+#[repr(C)]
+pub struct DatabaseCreationResult {
+    pub db: *const DatabaseHandle<'static>,
+    pub error_str: *const u8,
+}
+
+impl From<Result<Db, String>> for DatabaseCreationResult {
+    fn from(result: Result<Db, String>) -> Self {
+        match result {
+            Ok(db) => DatabaseCreationResult {
+                db: Box::into_raw(Box::new(db.into())),
+                error_str: std::ptr::null(),
+            },
+            Err(error_msg) => {
+                let error_cstring = CString::new(error_msg).unwrap_or_default().into_raw();
+                DatabaseCreationResult {
+                    db: std::ptr::null(),
+                    error_str: error_cstring.cast::<u8>(),
+                }
+            }
+        }
+    }
+}
+
+/// Frees the memory associated with a `DatabaseCreationResult`.
+/// This only needs to be called if the `error_str` field is non-null.
+///
+/// # Arguments
+///
+/// * `result` - The `DatabaseCreationResult` to free, previously returned from `fwd_create_db` or `fwd_open_db`.
+///
+/// # Safety
+///
+/// This function is unsafe because it dereferences raw pointers.
+/// The caller must ensure that `result` is a valid pointer.
+///
+/// # Panics
+///
+/// This function panics if `result` is `null`.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn fwd_free_database_error_result(result: *const DatabaseCreationResult) {
```

Comment:
This is undefined behavior. We are mutating what is owned by a const pointer. This should be a mut pointer instead. Or even plainly `&mut DatabaseCreationResult`.

```suggestion
pub unsafe extern "C" fn fwd_free_database_error_result(result: &mut DatabaseCreationResult) {
```

Reply:
Good catch! This occurred somewhere to, so I fixed the UB there as well.